WabiSabiMonitor.ApplicationCore/RoundStateStoreManager.cs

```diff
@@ -0,0 +1,79 @@
+
+namespace WabiSabiMonitor.ApplicationCore
+{
+    public class RoundStateStoreManager : BackgroundService
+    {
+        private readonly IRoundDataReaderService _roundDataReaderService;
+
+        public RoundStateStoreManager(IRoundDataReaderService roundDataReaderService)
+        {
+            _roundDataReaderService = roundDataReaderService;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            List<KeyValuePair<uint256, RoundDataReaderService.ProcessedRound>> savedForNextDay = new();
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var date = DateTime.Now.Date;
+                await WaitUntilMidnightAsync(stoppingToken);
+                Logger.LogInfo("It's midnight. Writing Daily Round States...");
+
+                Dictionary<uint256, RoundDataReaderService.ProcessedRound> allRounds = _roundDataReaderService.Rounds;
+
+                // Get rounds that happened today.
+                var roundsToday = allRounds.Where(x => x.Value.LastUpdate.Date == date).ToList();
+
+                // Add leftovers from yesterday.
+                roundsToday.AddRange(savedForNextDay);
+                savedForNextDay.Clear();
+
+                // Save finished rounds.
+                var finishedRounds = roundsToday.Where(x => x.Value.Round.EndRoundState != EndRoundState.None);
+
+                // If a BlameRound is still active, don't save its BlameOf.
+                var stillActiveBlameRoundsBlameOfIds = roundsToday.Where(x => x.Value.Round.EndRoundState == EndRoundState.None && x.Value.Round.IsBlame())
+                    .Select(x => x.Value.Round.BlameOf);
+
+                var dataToWrite = finishedRounds.Where(round =>
+                {
+                    bool isBlameOf = stillActiveBlameRoundsBlameOfIds.Contains(round.Key);
+                    if (isBlameOf)
+                    {
+                        Logger.LogInfo($"Excluded Round {round.Key}, is BlameOf.");
+                        savedForNextDay.Add(round);
+                    }
+
+                    return !isBlameOf;
+                });
+
+                var path = Path.Combine(EnvironmentHelpers.GetDataDir(Path.Combine("WabiSabiMonitor", "Client")), $"RoundStates_{date:yyyy-MM-dd}.json");
+                await File.WriteAllTextAsync(path, JsonConvert.SerializeObject(dataToWrite, JsonSerializationOptions.CurrentSettings), stoppingToken);
+            }
+        }
+
+        private async Task WaitUntilMidnightAsync(CancellationToken stoppingToken)
+        {
+            var now = DateTime.Now.TimeOfDay;
+            TimeSpan midnight = new(0, 0, 0);
+
+            if (now > midnight)
+            {
+                midnight = midnight.Add(TimeSpan.FromDays(1));
+            }
+
+            var waitUntil = midnight - now;
+
+            await Task.Delay(waitUntil, stoppingToken);
+        }
+    }
+}
```

Comment:
Instead of this, we could you make this class a 5 min periodic runner, save the day of the last time we saved the file, and at each time we run ExecuteAsync return if we're still at the same day. Please evaluate your solution and this one and take the decision what to do.

Reply:
I like my approach more as it doesn't require the thread to wake up every 5 minutes and check the time, instead it wakes up only once, do the work and goes to sleep again. We're talking about low amount of resources in both case, but it's still less in this one. I don't have any hard argument against yours either, but I'd go with mine.